### PR TITLE
Improved input syntax for `frequency_differencing` operation

### PR DESCRIPTION
**Pull request by praneethratna**
Addresses #967. Now input for frequency_differencing can be given directly in the form of frequency differencing criteria instead of separating out freq/chan from operators.

CC @leewujung

### CODE REVIEW

**Comment by emiliom** [comment]
These improvements are very cool! 

Since you're adding `sympy` as a dependency, how "heavy" a library is it? In other words, is it likely to create dependency resolution issues? I see the dependency tree `sympy > mpmath > gmpy2`, where [gmpy2](https://github.com/aleaxit/gmpy) apparently uses a C library ("gmpy2 is an optimized, C-coded Python extension module  ...").

Are there lighter-weight alternatives to `sympy` that would meet our needs? sympy has tons of capabilities, and here I think we'd be using only a tiny subset.

File: requirements-dev.txt [hunk 1]
```diff
@@ -23,3 +23,4 @@ sphinx_rtd_theme
 sphinxcontrib-mermaid
 twine
 wheel
+sympy
```
Comment (emiliom) [inline requirements-dev.txt:26]: I don't think sympy needs to go here, too. It only needs to be in `requirements.txt`
```suggestion
```

**Comment by praneethratna** [comment]
> These improvements are very cool! 
> 
> Since you're adding `sympy` as a dependency, how "heavy" a library is it? In other words, is it likely to create dependency resolution issues? I see the dependency tree `sympy > mpmath > gmpy2`, where [gmpy2](https://github.com/aleaxit/gmpy) apparently uses a C library ("gmpy2 is an optimized, C-coded Python extension module  ...").

I feel `SymPy` is lightweight compated compared to other mathematical libraries found online. Also `SymPy` has a hard dependency only on mpmath but I don't it would create dependency resolution issues.

> 
> Are there lighter-weight alternatives to `sympy` that would meet our needs? sympy has tons of capabilities, and here I think we'd be using only a tiny subset.

I'm using `SymPy` for parsing numbers from strings using it's inbuilt parser. I'm not sure of any other libraries which can be used for this.

**Comment by emiliom** [comment]
> Also SymPy has a hard dependency only on mpmath but I don't it would create dependency resolution issues.

Right! Scratch my comment about dependency resolution per se. I did see that the dependency was limited to mpmath > gmpy2, and we don't currently use those libraries. My apologies.

**Comment by praneethratna** [comment]
> > Also SymPy has a hard dependency only on mpmath but I don't it would create dependency resolution issues.
> 
> Right! Scratch my comment about dependency resolution per se. I did see that the dependency was limited to mpmath > gmpy2, and we don't currently use those libraries. My apologies.

No worries! I understand the clarification now. Since we don't use mpmath or gmpy2 libraries, these dependencies doesn't affect us. But we need to look at alternatives for `SymPy` if that makes the package slow.

**Comment by leewujung** [comment]
Since the equations are pretty predictable, one option would be to just do `re` all the way?

**Comment by praneethratna** [comment]
> Since the equations are pretty predictable, one option would be to just do `re` all the way?

You mean we use a single regex expression to check for the validity of the whole frequency differencing criteria? Wouldn't that make the code a bit messy?

**Comment by leewujung** [comment]
> @leewujung Do you mean we use a single regex expression to check for the validity of the whole frequency differencing criteria? Wouldn't that make the code a bit messy?

I think it would be ok? For the freqAB case, the `Hz/kHz/MHz/GHz` cases are fixed, so the number in front of that could be extracted. For the chanAB case, it is the string between the double quotes. The binary comparison operator `>, <, >=, <=, ==` are the allowed cases. And then we need to extracted the number before `dB`. This doesn't seem overly complicated?

**Comment by praneethratna** [comment]
> I think it would be ok? For the freqAB case, the `Hz/kHz/MHz/GHz` cases are fixed, so the number in front of that could be extracted. For the chanAB case, it is the string between the double quotes. The binary comparison operator `>, <, >=, <=, ==` are the allowed cases. And then we need to extracted the number before `dB`. This doesn't seem overly complicated?

@leewujung I think it shouldn't be overly complicated, it's just that i split everything into individual parts for better code debugging. I'll make these changes by today.

**Comment by praneethratna** [comment]
@leewujung Even if use a single `re` expression to check for validity, we again need to use separate `re`' expressions to extract freq/chan values, operators and diff or am i understanding your point in a wrong way?

**Comment by praneethratna** [comment]
@leewujung how do you suggest to do this?

**Comment by leewujung** [comment]
Hi @praneethratna : sorry for my super late response. What I am thinking in terms of using `re` to grab out the pattern is like the following:
- use patterns like `"(?P<freqA>\d+)(?P<unitA>\w?)Hz"` to match freqA and freqB, for example:
  ```python
  a = re.match("(?P<freq>\d+)\s*(?P<unit>\w?)Hz", "120kHz") # also matches arbitrary space between "120" and "kHz"
  a["freq"]  # get the number associated with the frequency
  a["unit"]  # get the unit associated with the frequency
  ```
  then you can use `a["unit"]` to determine the right multiplication factor to get to the numbers in `frequency_nominal`, which is with units Hz (i.e. 120kHz is 120000 in `frequency_nominal`).
- you can use the pattern above to build up the whole pattern, something like the following:
  ```python
  freqA_pattern = "(?P<freqA>\d+)\s*(?P<unitA>\w?)Hz"
  freqA_pattern = "(?P<freqB>\d+)\s*(?P<unitB>\w?)Hz"
  db_pattern = "(?P<db>\d+)\s*dB"
  cmp_pattern = "\s*(?P<cmp>\S+?)\s*"
  DIFF_MATCHER = re.compile(freqA_pattern + "\s*-\s*" + freqB_pattern + cmp_pattern + db_pattern)
  ```
  and then use outputs of the matched results with some error checking and handling (like converting `str` to `float`, k to 1000 etc) to combine with your other parts of the code.

What do you think?

**Comment by praneethratna** [comment]
> Hi @praneethratna : sorry for my super late response. What I am thinking in terms of using `re` to grab out the pattern is like the following:
> 
> * use patterns like `"(?P<freqA>\d+)(?P<unitA>\w?)Hz"` to match freqA and freqB, for example:
>   ```python
>   a = re.match("(?P<freq>\d+)\s*(?P<unit>\w?)Hz", "120kHz") # also matches arbitrary space between "120" and "kHz"
>   a["freq"]  # get the number associated with the frequency
>   a["unit"]  # get the unit associated with the frequency
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   then you can use `a["unit"]` to determine the right multiplication factor to get to the numbers in `frequency_nominal`, which is with units Hz (i.e. 120kHz is 120000 in `frequency_nominal`).
> * you can use the pattern above to build up the whole pattern, something like the following:
>   ```python
>   freqA_pattern = "(?P<freqA>\d+)\s*(?P<unitA>\w?)Hz"
>   freqA_pattern = "(?P<freqB>\d+)\s*(?P<unitB>\w?)Hz"
>   db_pattern = "(?P<db>\d+)\s*dB"
>   cmp_pattern = "\s*(?P<cmp>\S+?)\s*"
>   DIFF_MATCHER = re.compile(freqA_pattern + "\s*-\s*" + freqB_pattern + cmp_pattern + db_pattern)
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   and then use outputs of the matched results with some error checking and handling (like converting `str` to `float`, k to 1000 etc) to combine with your other parts of the code.
> 
> What do you think?

@leewujung No worries! I understand it now, these suggestions helped me to refactor the code in a much simpler and clean way. I have made some assumptions while commiting the changes (related to `frequency_nominal`) and the test code which generates `frequency_nominal` which might require a change before we merge this.

**Comment by leewujung** [comment]
@praneethratna : there are still places where you use `sympy` in the code. Could you remove them completely and use `re` instead? Thanks!

**Comment by praneethratna** [comment]
@leewujung I have removed the usage of `sympy' completely and also squashed all the commits into one.

File: echopype/mask/api.py [hunk 2]
```diff
@@ -366,58 +367,85 @@ def get_ch_shape(da):
...
+        If `freqABEq` or `chanABEq` is not a valid equation.
     """
 
-    # check that either freqAB or chanAB are provided and they are a list of length 2
-    if (freqAB is None) and (chanAB is None):
+    if (freqABEq is None) and (chanABEq is None):
         raise ValueError("Either freqAB or chanAB must be given!")
-    elif (freqAB is not None) and (chanAB is not None):
+    elif (freqABEq is not None) and (chanABEq is not None):
         raise ValueError("Only freqAB or chanAB must be given, but not both!")
```
Comment (leewujung) [inline echopype/mask/api.py:407]: minor wording tweak
```suggestion
    elif (freqABEq is not None) and (chanABEq is not None):
        raise ValueError("Only one of freqAB or chanAB should be given, but not both!")
```

**Review by leewujung — COMMENTED** [review]
@praneethratna : Thanks for this great PR!! It is awesome to have this intuitive way of putting in the freq diff equations.

I have some small suggestions for the organization and the tests:
- Since in #1142 there are more functions added to the `mask` subpackage, how about in this PR moving the frequency-differencing-related functions (`_check_source_Sv_freq_diff` and `check_freq_diff_and_parse_arguments`) to a separate module, say `echopype/mask/freq_diff.py` and keep the main API function `frequency_differencing` in `echopype/mask/api.py`? 
- for the tests, currently `test_frequency_differencing` is more of an integrated test, and in it the frequency units are all Hz so the unit matching was not tested. How about adding another set of tests to test for the function you wrote `check_freq_diff_and_parse_arguments`? This way you can test parsed element values and xfail the ones that should fail, etc.
- I suggest renaming `check_freq_diff_and_parse_arguments` to `_parse_freq_diff_eq` for the sake of clarity and also making it shorter

**Comment by leewujung** [comment]
Also, I caught a bug / undesired behavior in the current `_check_source_Sv_freq_diff` function. In the current implementation, the values of `frequency_nominal` is not allowed to be duplicated:
https://github.com/OSOceanAcoustics/echopype/blob/254ee106eb580959ca85f2131168c4f17c0a285a/echopype/mask/api.py#L463-L467

However, it is possible that there are 2 channels (with different `channel` strings) that have the same `frequency_nominal` values (this is actually why we decided to use `channel` as the coordinate, instead of `frequency_nominal`, since coordinates cannot have duplicated values), and users may want to diff these 2 channels using the `chanABEq`.  I think that potentially duplicated `frequency_nominal` values should only be checked if people use `freqABEq` to put in their equation.

Looking at the code again, what do you think if we take the following approach to change this function?
- keep the first checks `# check that channel and frequency nominal are in source_Sv` intact
- for the below three checks, make it such that only `freqAB` checks are executed when people use `freqABEq`, and only `chanAB` checks are executed when people use `chanABEq`
  ```
  # make sure that the channel and frequency_nominal values are not repeated in source_Sv
  # check that the elements of freqAB are in frequency_nominal
  # check that the elements of chanAB are in channel
  ```
   ie we only check for `frequency_nominal` to not be duplicated and elements of `freqAB` are in `frequency_nominal` when people use `freqABEq`, and only check for `channel` to not be duplicated and elements of `chanAB` are in `channel` when people use `chanABEq`.

File: echopype/tests/mask/test_mask.py [hunk 3]
```diff
@@ -287,33 +291,84 @@ def test_check_source_Sv_freq_diff(n: int, n_chan_freq: int, add_chan: bool, add
...
-        (5, 4, [1.0, 3.0], None, 1.0, ">", np.zeros((5, 5))),
-        (5, 4, None, ['chan1', 'chan3'], 1.0, ">", np.zeros((5, 5))),
-        (5, 4, [1.0, 3.0], None, 1.0, "<=", np.ones((5, 5))),
-        (5, 4, None, ['chan1', 'chan3'], 1.0, "<=", np.ones((5, 5))),
-        (5, 4, [1.0, 3.0], None, 1.0, "<", np.ones((5, 5)) - np.identity(5)),
-        (5, 4, None, ['chan1', 'chan3'], 1.0, "<", np.ones((5, 5)) - np.identity(5)),
+        ("1.0Hz-2.0Hz==1.0dB", None),
+        (None, '"chan1"-"chan3"==1.0dB'),
+        ("1.0 kHz - 2.0 MHz>=1.0 dB", None),
+        (None, '"chan2-12 89" - "chan4 89-12" >= 1.0 dB'),
```
Comment (leewujung) [inline echopype/tests/mask/test_mask.py:303]: These variations are great! I can see that the current test makes sure that `_parse_freq_diff_eq` is run through without problem, though the values are not checked. Could you add a section in the test for the parsed values, to make sure the output is correct? Thanks!

File: echopype/mask/freq_diff.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            "The provided list input freqAB contains values that "
+            "are not in the frequency_nominal variable!"
+        )
+
+    # check that the elements of chanAB are in channel
+    if (chanAB is not None) and (not all([chan in source_Sv.channel for chan in chanAB])):
+        raise ValueError(
+            "The provided list input chanAB contains values that are "
+            "not in the channel coordinate!"
+        )
```
Comment (leewujung) [inline echopype/mask/freq_diff.py:148]: I would suggest consolidating the above to improve readability:
```python
# raise error if freqAB and chanAB are both None or both have values
if ...
   raise ValueError("...")

if freqAB is not None:
   # test if frequency_nominal values are not repeated in source_Sv
   # check that the elements of freqAB are in frequency_nominal
else:  # this implies chanAB is not None
   # test if channel values are not repeated in source_Sv
   # check that the elements of chanAB are in frequency_nominal
```

File: echopype/tests/mask/test_mask.py [hunk 5]
```diff
@@ -232,21 +236,21 @@ def create_input_mask(
...
                      marks=pytest.mark.xfail(strict=True,
                                              reason="This should fail because the Dataset "
                                                     "will not have the frequency_nominal variable.")),
-        pytest.param(5, 3, True, True, [1.0, 4.0], None,
+        pytest.param(5, 4, True, True, [1.0, 4.0], None,
                      marks=pytest.mark.xfail(strict=True,
                                              reason="This should fail because not all selected frequencies"
                                                     "are in the frequency_nominal variable.")),
-        pytest.param(5, 3, True, True, None, ['chan1', 'chan4'],
+        pytest.param(5, 4, True, True, None, ['chan1', 'chan9'],
```
Comment (leewujung) [inline echopype/tests/mask/test_mask.py:253]: I'd suggest another 2 cases where you xfail duplicated `freqAB` and duplicated `chanAB` cases, just to be complete.

**Review by leewujung — CHANGES_REQUESTED** [review]
Thanks @praneethratna for your speedy changes! I added some suggestions, and I think once they are addressed we can merge this PR. :)

**Review by leewujung — APPROVED** [review]
@praneethratna : Thanks a lot for your contribution. We can merge this PR now!

**Comment by leewujung** [comment]
Ugh sorry just remember one thing. In here
https://github.com/OSOceanAcoustics/echopype/blob/254ee106eb580959ca85f2131168c4f17c0a285a/echopype/tests/utils/test_processinglevels_integration.py#L106
`frequency_differencing` is called (this is something the CI does not catch, unless one uses [all tests ci] (see [here](https://echopype.readthedocs.io/en/latest/contributing.html#continuous-integration-github-actions)). Could you quickly change it? Then I will merge. Thank you!

**Pull request merged**